### PR TITLE
resolved: filter stub listeners in manager_get_dns_server()

### LINKED ISSUE #20555: [resolved] Servers in DNSStubListenerExtra= not filtered out
**systemd version the issue has been seen with**
 > 249.3

**Used distribution**
 > Arch Linux

**Linux kernel version used** (`uname -a`)

 > `Linux ideapad 5.13.12-arch1-1 #1 SMP PREEMPT Wed, 18 Aug 2021 20:49:03 +0000 x86_64 GNU/Linux`

**CPU architecture issue was seen on**
 > x86_64

**Expected behaviour you didn't see**
 > Address(es) listed in `DNSStubListenerExtra=` works as fine as `127.0.0.53`

**Unexpected behaviour you saw**
 When I have e.g. `nameserver 192.168.150.1` (in addition to, regardless of the order / replacing `nameserver 127.0.0.53` ) in `/etc/resolv.conf`, I will get `SERVFAIL` error when I `nslookup`. The problem requires two additional conditions to occur though:

1. `/etc/resolv.conf` is a regular file (instead of a symlink to `/run/systemd/resolve/stub-resolv.conf`)
2. `search .` is used

**Steps to reproduce the problem**
 ```
# realpath /etc/resolv.conf 
/run/systemd/resolve/stub-resolv.conf
# grep -v '\(^#\|\(^$\)\)' /etc/resolv.conf 
nameserver 192.168.150.1
options edns0 trust-ad
search .
# nslookup github.com
Server:		192.168.150.1
Address:	192.168.150.1#53

Non-authoritative answer:
Name:	github.com
Address: 52.74.223.119

# rm /etc/resolv.conf 
# grep -v '\(^#\|\(^$\)\)' /run/systemd/resolve/stub-resolv.conf > /etc/resolv.conf
# nslookup github.com
Server:		192.168.150.1
Address:	192.168.150.1#53

** server can't find github.com: SERVFAIL
```
```
# cat /etc/resolv.conf 
nameserver 192.168.150.1
options edns0 trust-ad
# nslookup github.com
Server:		192.168.150.1
Address:	192.168.150.1#53

Non-authoritative answer:
Name:	github.com
Address: 20.205.243.166

```

### PR DESCRIPTION
**Pull request by tomty89**
Fixes #20555 and replacing #20556. ~There are probably some more left to be cleaned up.~ Let me know if you find the approach alright, thanks! @poettering

### CODE REVIEW

**Comment by tomty89** [comment]
Now it's all set

**Comment by tomty89** [comment]
@poettering I wonder if the stub-loop guard should still be preserved for transactions that are of a link scope. It's definitely not the right way to handle things when `!t->scope-link` though.

**Comment by tomty89** [comment]
Ping @poettering

**Comment by keszybz** [comment]
The commit is lacking a description. Please describe the problem that is being fixed, and how the previous solution didn't work, and how this one fixes the issue.

**Comment by tomty89** [comment]
I am not sure if it has ever been completely clear to me `how the previous solution didn't work`. (Actually, what's unclear to me is perhaps `how the previous solution could have worked in some case` instead; I think I dug something but I might have forgotten to jot it down). Also see #20555.

Anyway though, the "-ELOOP approach" does not exactly make sense to me:
1. Given the scenario that an extra stub listener is configured and added as the SOLE nameserver to /etc/resolv.conf
2. The extra stub listener becomes the only nameserver of type `DNS_SERVER_SYSTEM`.
3. `dns_scope_next_dns_server(t->scope, t->server)` that is used to deal with the `-ELOOP` calls `manager_next_dns_server()`
4. `(dns_scope_get_dns_server(t->scope) == t->server)` is doomed to be true in such scenario because of how `manager_next_dns_server()` works / is expected to work

The expected / desired heuristic in such scenario is not different from the usual:
1.`current_dns_server` set to a fallback one if there there's not any link nameserver
2. otherwise, `current_dns_server` set to / remains being NULL.

Therefore I get `manager_get_dns_server()` to do its / a better job:
1. Check (if set) whether `current_dns_server` is a stub before returning it
2. If so, see if there's another nameserver of type `DNS_SERVER_SYSTEM`
3. Repeat step 1 and 2 until one is found OR there's no more to check
4. If all nameservers of type `DNS_SERVER_SYSTEM` are stub, set current_dns_server to NULL (i.e. cause everything to behave as if there's not any nameserver of `DNS_SERVER_SYSTEM` type)

Putting it in another way: the reason to do so is that when it's not of scope link, it's not actually a dead end when all we have (from the list that consists of `DNS_SERVER_SYSTEM` servers) is a stub, so there's no reason to simply complete the transaction (or "forcing" that to happen) with the reason of "stub loop".

**Comment by tomty89** [comment]
@poettering @keszybz I've updated the commit message (which should explains it even better than my last comment) and withdrawn the (partial) reversion of 49ef064. I hope that this can finally gets in.

File: src/resolve/resolved-dns-server.c [hunk 1]
```diff
@@ -878,6 +878,13 @@ DnsServer *manager_get_dns_server(Manager *m) {
         if (!m->current_dns_server)
                 manager_set_dns_server(m, m->dns_servers);
 
+        while (m->current_dns_server &&
+               manager_server_is_stub(m, m->current_dns_server)) {
+                manager_next_dns_server(m, NULL);
+                if (m->current_dns_server == m->dns_servers)
```
Comment (tomty89) [inline src/resolve/resolved-dns-server.c:884]: I'm not sure if this is good enough btw. I mean, is it possible that when the loop began, `current_dns_server` is not `dns_servers` (the beginning of the list)?

I thought of remembering the `current_dns_server` before loop begins, but it seems like a server could be "unlinked from the list"? In which case the loop will become infinite?

File: src/resolve/resolved-dns-server.c [hunk 2]
```diff
@@ -878,6 +878,13 @@ DnsServer *manager_get_dns_server(Manager *m) {
         if (!m->current_dns_server)
```
Comment (tomty89) [inline src/resolve/resolved-dns-server.c:878]: Maybe I should add `|| manager_server_is_stub(m, m->current_dns_server)` here?

File: src/resolve/resolved-dns-server.c (hunk 1 above)
Comment (tomty89) [reply src/resolve/resolved-dns-server.c:884]: Maybe I should add `|| manager_server_is_stub(m, m->current_dns_server)` here?

File: src/resolve/resolved-dns-server.c (hunk 2 above)
Comment (tomty89) [inline src/resolve/resolved-dns-server.c:878]: Maybe I should add `|| manager_server_is_stub(m, m->current_dns_server)` here?

**Comment by keszybz** [comment]
Hmm, this seems to make things work, but I think resolved is still confused.

I created the following setup:
```
$ ls -l /etc/resolv.conf
-rw-r--r-- 1 root root 57 Dec 15 10:26 /etc/resolv.conf
$ cat /etc/resolv.conf
nameserver 192.168.150.1
options edns0 trust-ad
search .
$ cat /etc/systemd/resolved.conf.d/stub.conf
[Resolve]
DNSStubListenerExtra=192.168.150.1
```
systemd-resolved listens on 192.168.150.1 as configured:
```console
$ sudo lsof -n -p $(systemctl -P MainPID show systemd-resolved)|grep 192.168.150
systemd-r 48494 systemd-resolve   22u     IPv4            2430811      0t0     UDP 192.168.150.1:domain 
systemd-r 48494 systemd-resolve   23u     IPv4            2430812      0t0     TCP 192.168.150.1:domain (LISTEN)
$ sudo ip a a 192.168.150.1/24 dev host0
$ dig wp.pl @192.168.150.1              
;; ANSWER SECTION:
wp.pl.			107	IN	A	212.77.98.9
```

So far so good. But:
```console
$ resolvectl               
Global
    resolv.conf mode: foreign
         DNS Servers: 192.168.150.1
```
I don't think this is useful. If we *know* that it's us listening on that address, we shouldn't treat this as an "upstream" server.

**Comment by keszybz** [comment]
I pushed a commit to filter out our own listeners earlier. Together with your fix, I think this should solve the issue nicely.

**Comment by tomty89** [comment]
Well while the stub will be listed in "DNS Servers", it will never become the "Current DNS Server". What you saw is expected. The approach has been chosen because it's more "generic". See #20556 for more details.

I have to admit that it's for sort of a corner case. It was a problem to me when I would like a private network created with network namespace to use the stub (which is listening in the main netns only) as DNS server. By having the stub in the resolv.conf (which is "shared" by both netns), it acts as a  "remote" DNS server and a stub respectively for the two netns.

To be fair, it's only works when there's only one such extra netns network. Probably the general solution is to use dnsmasq (which listens to 127.0.0.53:53 and forward request to corresponding extra stub address) in each netns anyway, which is what I've been doing so far. Still it doesn't seem bad to tie up such loose end in systemd.

**Comment by keszybz** [comment]
Well, OK… "more generic". But do you actually think the change I added is OK or not?

**Comment by tomty89** [comment]
I think what it does is good, but I do have a bit of concern that now we have like three layers (the `-ELOOP`, the get filter, the add filter) of "guards" to deal with essentially just one issue. I feel like that could bring some maintenance cost (especially for newcomers), yet what the add filter fixes is now somewhat cosmetic.

Besides personally I don't like inconsistent behavior (although there seem to be reasons, which I don't grok, that you think we should keep the bus part unguarded).

While I'm not entirely comfortable with it, I don't strongly oppose it either. The decision is all yours.

**Comment by poettering** [comment]
Honestly I am not convinced the filtering really is desirable. I think it should be OK to advertise extra stub listeners to local clients, but you prohibit this now. i.e. there are two different concepts here, and we shouldn't mix them up:

1. tracking configured dns servers and advertise them to local programs
2. actually using them ourselves

I am pretty sure that our own stubs are OK for 1 but not OK for 2, hence we should filter at the time of *use* not at the time of *parse*.

Thus I think ew should revert the last commit in this PR, in particular as it's broken anyway, see above.

**Pull request merged**